core/chains/evm/logpoller/orm.go

```diff
@@ -0,0 +1,111 @@
+	q := o.q.WithOpts(qopts...)
+	var b LogPollerBlock
+	if err := q.Get(&b, `SELECT * FROM log_poller_blocks WHERE block_hash = $1 AND evm_chain_id = $2`, h, utils.NewBig(o.chainID)); err != nil {
+		return nil, err
+	}
+	return &b, nil
+}
+
+func (o *ORM) SelectBlockByNumber(n int64, qopts ...pg.QOpt) (*LogPollerBlock, error) {
+	q := o.q.WithOpts(qopts...)
+	var b LogPollerBlock
+	if err := q.Get(&b, `SELECT * FROM log_poller_blocks WHERE block_number = $1 AND evm_chain_id = $2`, n, utils.NewBig(o.chainID)); err != nil {
+		return nil, err
+	}
+	return &b, nil
+}
+
+func (o *ORM) SelectLatestBlock(qopts ...pg.QOpt) (*LogPollerBlock, error) {
+	q := o.q.WithOpts(qopts...)
+	var b LogPollerBlock
+	if err := q.Get(&b, `SELECT * FROM log_poller_blocks WHERE block_number = (SELECT MAX(block_number) FROM log_poller_blocks) AND evm_chain_id = $1`, utils.NewBig(o.chainID)); err != nil {
+		return nil, err
+	}
+	return &b, nil
+}
+
+func (o *ORM) DeleteRangeBlocks(start, end int64, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+	_, err := q.Exec(`DELETE FROM log_poller_blocks WHERE block_number >= $1 AND block_number <= $2 AND evm_chain_id = $3`, start, end, utils.NewBig(o.chainID))
+	return err
+}
+
+func (o *ORM) DeleteLogs(start, end int64, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+	_, err := q.Exec(`DELETE FROM logs WHERE block_number >= $1 AND block_number <= $2 AND evm_chain_id = $3`, start, end, utils.NewBig(o.chainID))
+	return err
+}
+
+// InsertLogs is idempotent to support replays.
+func (o *ORM) InsertLogs(logs []Log, qopts ...pg.QOpt) error {
+	q := o.q.WithOpts(qopts...)
+	_, err := q.NamedExec(`INSERT INTO logs 
+(evm_chain_id, log_index, block_hash, block_number, address, event_sig, topics, tx_hash, data, created_at) VALUES 
+(:evm_chain_id, :log_index, :block_hash, :block_number, :address, :event_sig, :topics, :tx_hash, :data, NOW()) ON CONFLICT DO NOTHING`, logs)
+	return err
+}
+
+func (o *ORM) selectLogsByBlockRange(start, end int64) ([]Log, error) {
+	var logs []Log
+	err := o.q.Select(&logs, `
+        SELECT * FROM logs 
+        WHERE block_number >= $1 AND block_number <= $2 AND evm_chain_id = $3
+        ORDER BY (block_number, log_index, created_at)`, start, end, utils.NewBig(o.chainID))
+	if err != nil {
+		return nil, err
+	}
+	return logs, nil
+}
+
+// SelectLogsByBlockRangeFilter finds the latest logs by block.
+// Assumes that logs inserted later for a given block are "more" canonical.
+func (o *ORM) SelectLogsByBlockRangeFilter(start, end int64, address common.Address, eventSig []byte, qopts ...pg.QOpt) ([]Log, error) {
+	var logs []Log
+	q := o.q.WithOpts(qopts...)
+	err := q.Select(&logs, `
+		SELECT * FROM logs 
+			WHERE logs.block_number >= $1 AND logs.block_number <= $2 AND logs.evm_chain_id = $3 
+			AND address = $4 AND event_sig = $5 
+			ORDER BY (logs.block_number, logs.log_index)`, start, end, utils.NewBig(o.chainID), address, eventSig)
```

Comment:
This looks like the order by will require an in mem sort after selecting from the index. Did you consider sorting on the node instead? Maybe not significant :shrug:

Reply:
yeah seems negligible